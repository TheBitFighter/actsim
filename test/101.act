

// compare
export template<pint w_in, w_out>
function sign_extend (int<w_in> in) : int<w_out>
{
    int<w_out> a,b;

  chp {
        // sign_extend alternative
        a := in{w_in-1}=1 ? {~int(0,w_out-w_in),in} : {int(0,w_out-w_in),in};
   
        // std sign extend
        [  in{w_in-1} = 1 -> b := in | (~int(0,w_out) << w_in)
        [] else -> b := in
        ];

        //compare them
        [  a!=b -> log(" input: ",in," DIFFERENT a: ",a," b: ",b,"%b a: ",a,"%b b: ",b )
        [] a=b  -> log(" input: ",in," SAME      a: ",a," b: ",b,"%b a: ",a,"%b b: ",b )
        ];		

        self:=a
  }
}


pint win=15;
pint wout=32;
pint examples=10;

defproc test()
{
    int<win> v[10];
    int<wout> out;
    int<10> i;

    chp{
    
    v[0]:=65468;
    v[1]:=65451;
    v[2]:=65500;
    v[3]:=65488;
    v[4]:=65526;
    v[5]:=65528;
    v[6]:=65485;
    v[7]:=65492;
    v[8]:=65481;
    v[9]:=65486;
    
    i:=0;
    *[ i< examples -> 
            out := sign_extend<win,wout> (v[i]);
            i:=i+1
    ]

    }
}
