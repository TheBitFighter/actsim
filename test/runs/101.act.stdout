[                 120] <>   input: 32700 SAME      a: 4294967228 b: 4294967228 a: 11111111111111111111111110111100 b: 11111111111111111111111110111100
[                 140] <>   input: 32683 SAME      a: 4294967211 b: 4294967211 a: 11111111111111111111111110101011 b: 11111111111111111111111110101011
[                 160] <>   input: 32732 SAME      a: 4294967260 b: 4294967260 a: 11111111111111111111111111011100 b: 11111111111111111111111111011100
[                 180] <>   input: 32720 SAME      a: 4294967248 b: 4294967248 a: 11111111111111111111111111010000 b: 11111111111111111111111111010000
[                 200] <>   input: 32758 SAME      a: 4294967286 b: 4294967286 a: 11111111111111111111111111110110 b: 11111111111111111111111111110110
[                 220] <>   input: 32760 SAME      a: 4294967288 b: 4294967288 a: 11111111111111111111111111111000 b: 11111111111111111111111111111000
[                 240] <>   input: 32717 SAME      a: 4294967245 b: 4294967245 a: 11111111111111111111111111001101 b: 11111111111111111111111111001101
[                 260] <>   input: 32724 SAME      a: 4294967252 b: 4294967252 a: 11111111111111111111111111010100 b: 11111111111111111111111111010100
[                 280] <>   input: 32713 SAME      a: 4294967241 b: 4294967241 a: 11111111111111111111111111001001 b: 11111111111111111111111111001001
[                 300] <>   input: 32718 SAME      a: 4294967246 b: 4294967246 a: 11111111111111111111111111001110 b: 11111111111111111111111111001110
